src/lib/pubkey/pk_options.h

```diff
@@ -0,0 +1,197 @@
+/*
+* (C) 2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_PK_OPTIONS_H_
+#define BOTAN_PK_OPTIONS_H_
+
+#include <botan/pk_keys.h>
+#include <optional>
+#include <span>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace Botan {
+
+/**
+* Signature generation/verification options
+*
+* The normal usage of this is in a builder style, eg
+*
+* PK_Signature_Options()
+*   .with_hash("SHA-256")
+*   .with_der_encoded_signature()
+*   .with_context("Foo")
+*/
+class BOTAN_PUBLIC_API(3, 6) PK_Signature_Options {
+   public:
+      /// Create an empty PK_Signature_Options
+      ///
+      /// This can be further parameterized by calling with_xxx functions
+      PK_Signature_Options() = default;
+
+      PK_Signature_Options(PK_Signature_Options&& other) = default;
+      PK_Signature_Options& operator=(PK_Signature_Options&& other) = default;
+
+      PK_Signature_Options(const PK_Signature_Options&) = delete;
+      PK_Signature_Options& operator=(const PK_Signature_Options& other) = delete;
+      ~PK_Signature_Options();
+
+      /// Format this PK_Signature_Options as a string
+      ///
+      /// This is primarily intended for debugging and error messages;
+      /// the format is not fixed
+      std::string to_string() const;
+
+      /// Create a PK_Signature_Options specifying the hash to use
+      ///
+      /// Most but not all signture schemes require specifying the hash
+      ///
+      /// This can be further parameterized by calling with_xxx functions
+      PK_Signature_Options(std::string_view hash_fn) : m_hash_fn(hash_fn) {}
+
+      /// Specify the hash function to use for signing/verification
+      ///
+      /// Most, but not all, schemes require specifying a hash function.
+      PK_Signature_Options with_hash(std::string_view hash) &&;
```

Comment:
I'm a bit surprised by the constraints on the builder methods. I would have expected:

```C++
PK_Signature_Options& with_hash(std::string_view hash) {
  set_or_throw(m_hash, hash);
  return *this;
}
```

Some users would probably want to express things like this:

```C++
auto some_key = get_private_key();

auto opts = PK_Signature_Options("SHA-256");
if(key->algo_name() == "RSA) {
   opts.with_padding("PSS");
}
if(key->algo_name() == "ML-DSA") {
  opts.with_context("some context");
}

PK_Signer signer(some_key, opts);
```

I think people will grumble and then go for `std::move(opts).with_context(...)` as a workaround. See `pk_options_impl.cpp` 😜.

Reply:
I went this way to avoid the problem that if there is a `unique_ptr<HashFunction>` in there we have to clone it for each and every `with_foo`. There is maybe a better way to handle this.

Reply:
Returning a `PK_Options&` (`return *this`) avoids the copy, as long as the user doesn't assign it to another variable. That allows chaining `with_*` just fine.

Reply:
I have to correct myself to a certain extent: When there's a `unique_ptr<HashFunction>` you are forced to first create an lvalue of the `PK_Options`. E.g. this wouldn't work:

```C++
auto opts = PK_Options().with_hash(Hash_Function::create_or_throw("SHA-256"));
// Above fails, because `with_hash()` returns a `PK_Options&` that cannot be copied into `auto opts`
// (because `PK_Options` contains a unique_ptr).

if (key->algo() == "ML-DSA") {
   opts.with_context("some context");
}
```

In C++20 we'd have to overload `with_hash()` for lvalue-ref and rvalue-ref. Providing both overloads, would compile the above example, as the rvalue-ref variant of `with_hash()` would be called on the temporary object and the temporary would just propagate through `with_hash()` and eventually end up initializing the `auto opts` lvalue.

```C++
class PK_Options {
   PK_Options& with_hash(std::unique_ptr<HashFunction> hash) & {
      m_hash = std::move(hash);
      return *this;
   }

   PK_Options with_hash(std::unique_ptr<HashFunction> hash) && {
      m_hash = std::move(hash);
      return std::move(*this);
   }
};
```

[C++23 introduces "deducing this"](https://devblogs.microsoft.com/cppblog/cpp23-deducing-this/) which would come in handy here. Essentially, `self` is deduced to either an lvalue-ref or an rvalue-ref. And the `decltype(auto)` would just pass the lvalue/rvalue-ref along. The below snippet would also compile the above example in C++23 😢

```C++
class PK_Options {
   decltype(auto) with_hash(this auto&& self, std::unique_ptr<HashFunction> hash) {
      self.m_hash = std::move(hash);
      return self;
   }
};
```